keps/sig-scheduling/5598-opportunistic-batching/README.md

```diff
@@ -0,0 +1,967 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes an opportunistic batching mechanism in the scheduler to improve performance of scheduling many compatible pods at once, and to begin building the infrastructure required for gang scheduling. To implement this mechanism we propose the following additions:
+
+ - **Pod scheduling signature:** A signature that captures the properties of a pod that impact scoring and feasibility.
+ - **Batching mechanism:** A mechanism to reuse the scheduling output from one pod to set the nominated node name for multiple subsequent pods with matching scheduling signatures.
+ - **Opportunistic batching:** Transparent inclusion of the batching mechanism in the scheduler to improve the performance of targeted workloads that could benefit from it.
+
+## Motivation
+
+Today our scheduling algorithm is O(num pods x num nodes). As the size of clusters and jobs continue to increase, this leads to low performance when scheduling or rescheduling large jobs. This increases user cost and slows down user jobs, both unpleasant impacts. Optimizations like this one have the potential to dramaticly reduce the cost of scheduling in these scenarios.
+
+We are also working on gang scheduilng (in addition to other forms of multi-pod scheduling), which will give us a way to consider multiple pods at the same time. "Opportunistic batching" provides a starting point for these mechanisms by providing signatures and batching, both necessary foundational mechanisms, and including them initially in a simple way.
+
+Another change is the shift towards 1-pod-per-node in batch and ML environments. Many of these environments (among others) only attempt to run a single user pod on each node, along with a complement of daemon set pods. This simplifies our scheduling needs significantly, as it allows to reuse not only filtering, but also scoring results.
+
+### Goals
+
+ * Improve the performance of scheduling large jobs on large clusters where the constraints are simple.
+ * Begin building infrastructure to support gang scheduling and other "multi-pod" scheduling mechanisms.
+ * Ensure that the infrastructure we build is maintainable as we update, add and remove plugins.
+ * Provide improved performance with identical results to our current scheduler for targeted workloads in this release.
+ * Provide a path where we can expand batching to apply to most or all workloads over the next few releases.
+
+### Non-Goals
+
+ * We are not attempting to apply this optimization to all pods in this release. We will make the addition of batching transparent, but only applicable to a reduced set of workloads in v1.
```

Comment:
nit: We usually don't call v1 for features. You should rephrase it to something like "not in alpha, but we'll support more in Beta (or GA)", or "not in this KEP, but we'll support more in another KEP"